src/software/geom/voronoi_util.cpp

```diff
@@ -0,0 +1,96 @@
+#include <g3log/loglevels.hpp>
+#include <g3log/g3log.hpp>
+#include "software/geom/voronoi_util.h"
+#include "software/geom/util.h"
+
+
+std::vector<Point> findVoronoiEdgeRecIntersects(const voronoi_diagram<double>& vd, Rectangle rec, std::vector<Point> points)
+{
+    std::vector<Point> intersects;
+
+    for (auto edge : vd.edges())
+    {
+        // Edges extending outside the rectangle will be infinite edges
+        if (!edge.is_finite() && edge.is_primary())
+        {
+            auto start = edge.vertex0();
+            if (start)
+            {
+                // The direction of the infinite vector will be perpendicular to the
+                // vector connecting the two points which own the two half edges. We can
+                // use this to calculate another point on the infinite edge.
+                Point p1    = points[edge.cell()->source_index()];
+                Point p2    = points[edge.twin()->cell()->source_index()];
+                double endX = (p1.y() - p2.y());
+                double endY = (p1.x() - p2.x()) * -1;
+                // Extend the edge out to beyond the rectangle to ensure interception
+                // functions work.
+                Point end = Point(endX, endY) * dist(rec.furthestCorner(p2), p2);
+
+                std::vector<Point> edgeIntersects = lineRectIntersect(rec, Point(start->x(), start->y()), end);
+
+                intersects.insert(intersects.end(), edgeIntersects.begin(), edgeIntersects.end());
+            }
+        }
+    }
+    return intersects;
+}
+
+std::vector<Circle> voronoiVerticesToOpenCircles(const voronoi_diagram<double>& vd, Rectangle rec, std::vector<Point> points)
```

Comment:
Perhaps change `rec` to `bounding_box` or similar? Same goes for other places where we're passing in the rectangle as a bounding box.